### PR TITLE
feat: Fail all tasks for a disconnected Task Runner (no-changelog)

### PR DESCRIPTION
**Pull request by valya**
## Summary

When a Task Runner disconnects, the Task Broker should kill all pending Tasks and Task Offers, so that the workflow fails gracefully.

## Review / Merge checklist

- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) 
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [x] Tests included. 
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

### CODE REVIEW

File: packages/cli/src/runners/task-broker.service.ts [hunk 1]
```diff
@@ -75,15 +75,11 @@ export class TaskBroker {
 
 	expireTasks() {
 		const now = process.hrtime.bigint();
-		const invalidOffers: number[] = [];
-		for (let i = 0; i < this.pendingTaskOffers.length; i++) {
-			if (this.pendingTaskOffers[i].validUntil < now) {
-				invalidOffers.push(i);
+		for (let i = this.pendingTaskOffers.length; i >= 0; i--) {
+			if (this.pendingTaskOffers[i]?.validUntil < now) {
+				this.pendingTaskOffers.splice(i, 1);
```
Comment (ivov) [inline packages/cli/src/runners/task-broker.service.ts:80]: I'd expect to reverse loop like this:

```suggestion
		for (let i = this.pendingTaskOffers.length - 1; i >= 0; i--) {
			if (this.pendingTaskOffers[i].validUntil < now) {
				this.pendingTaskOffers.splice(i, 1);
```

File: packages/cli/src/runners/task-broker.service.ts [hunk 2]
```diff
@@ -92,6 +88,21 @@ export class TaskBroker {
 
 	deregisterRunner(runnerId: string) {
 		this.knownRunners.delete(runnerId);
+
+		// Remove any pending offers
+		for (let i = this.pendingTaskOffers.length; i >= 0; i--) {
+			if (this.pendingTaskOffers[i]?.runnerId === runnerId) {
+				this.pendingTaskOffers.splice(i, 1);
+			}
+		}
```
Comment (ivov) [inline packages/cli/src/runners/task-broker.service.ts:97]: See above

File: packages/cli/src/runners/task-broker.service.ts [hunk 3]
```diff
@@ -92,6 +88,21 @@ export class TaskBroker {
...
+
+		// Fail any tasks
+		for (const task of this.tasks.values()) {
+			if (task.runnerId === runnerId) {
+				void this.failTask(task.id, 'The Task Runner has disconnected');
+				this.handleRunnerReject(task.id, 'The Task Runner has disconnected');
```
Comment (ivov) [inline packages/cli/src/runners/task-broker.service.ts:103]: We could add the runner ID to the error message to help debugging.

File: packages/cli/src/runners/__tests__/task-broker.test.ts [hunk 4]
```diff
@@ -102,6 +102,55 @@ describe('TaskBroker', () => {
...
+			taskBroker.registerRunner(runner, messageCallback);
+			taskBroker.taskOffered({
+				offerId: 'offer1',
+				runnerId,
+				taskType: 'mock',
+				validFor: 1000,
+				validUntil: createValidUntil(1000),
+			});
+			taskBroker.taskOffered({
+				offerId: 'offer1',
```
Comment (ivov) [inline packages/cli/src/runners/__tests__/task-broker.test.ts:120]: It'd be clearer to use distinct offer IDs.

File: packages/cli/src/runners/__tests__/task-broker.test.ts [hunk 5]
```diff
@@ -102,6 +102,55 @@ describe('TaskBroker', () => {
...
+				runnerId: 'runner2',
+				taskType: 'mock',
+				validFor: 1000,
+				validUntil: createValidUntil(1000),
+			});
+			taskBroker.deregisterRunner(runnerId);
+
+			const offers = taskBroker.getPendingTaskOffers();
+			expect(offers).toHaveLength(1);
+			expect(offers.filter((o) => o.runnerId !== runnerId)).toHaveLength(1);
```
Comment (ivov) [inline packages/cli/src/runners/__tests__/task-broker.test.ts:130]: This assertion could be more explicit, i.e. assert that the remaining offer is for `runner2`.

**Review by ivov — APPROVED** [review]

**Comment by janober** [comment]
Got released with [`n8n@1.64.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.64.0)

**Pull request merged**